dartdoc_test/lib/src/logger.dart

```diff
@@ -0,0 +1,26 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:io';
+
+var verbose = false;
+
+enum LogLevel { debug, info, warning, error }
+
+void log(String message, [LogLevel level = LogLevel.info]) {
+  if (level == LogLevel.debug && !verbose) {
+    return;
+  }
+  stdout.writeln(message);
+}
```

Comment:
Instead of making global state like this, I'd suggest:
```
typedef Logger = void Function(String message, [LogLevel level = LogLevel.info]);
```

Then when you create an instance of `DartdocTest` you pass it a `Logger` in the constructor.

That way `AnalyzeCommand` will create a callback that it can pass in as a logger.
And then `DartdocTest` doesn't need to know whether or not it's running in verbose mode or not. It just writes log messages to a `Logger` function. And `AnalyzeCommand` just creates a logger function that prints what it wants to print based on the verbosity argument.

----------------------

## If we want to be a bit more ergonomic

To be fair, I might suggest making it a little fancier. Like create a class like:
```dart
// In dartdoc_test/lib/src/logger.dart

enum LogLevel { debug, info, warning, error }
typedef LoggerFn = void Function(LogLevel level, String message);

final class Logger {
  final LoggerFn _log;
  Logger(this._log);

  void debug(String message) => log(LogLevel.debug, message);
  void info(String message) => log(LogLevel.info, message);
  void warning(String message) => log(LogLevel.warning, message);
  void error(String message) => log(LogLevel.error, message);
  void log(LogLevel level, String message) => _log(level, message);
}

// In dartdoc_test/lib/src/command.dart

class AnalyzeCommand extends Command<void> {
  @override
  String get name => 'analyze';

  @override
  String get description => 'Analyze code samples in the current directory.';

  AnalyzeCommand() {
    argParser.addFlag(
      'write',
      abbr: 'w',
      help: 'Write code samples to file.',
    );
  }

  @override
  Future<void> run() async {
    final verbose = globalResults?.flag('verbose') ?? false;
    final logger = Logger((level, message) {
      if (verbose || LogLevel.debug != level) {
        stdout.writeln(message);
      }
    });

    final dartdocTest = DartdocTest(logger, DartdocTestOptions(
      write: argResults?.flag('write') ?? false,
    ));
    await dartdocTest.analyze();
  }
}

```

This way when you log message you don't have to pass `LogLevel`, you just do `logger.debug(...)` or `logger.info(..)`.

-------------------

## We could also be even more high-level. If you look at `package:test` it's possible to do `dart test --reporter expanded` or `dart test --reporter json`, there is even a `dart test --reporter github` that prints output with nice unicode characters that shows up as checkmarks on github.

So perhaps instead of writing everything as log messages to a `Logger`, perhaps we should just pass the results to a `Reporter` object. Then the `Reporter` is responsible for printing the output to stdout (or write it to a file). And we can put in different reporter objects depending on what kind of output we want.


```
enum Severity {
  info,
  warning,
  error,
}

/// Reporter to which [DartdocTest] will report results.
///
/// DartdocTest will report results to a [Reporter], when doing so it will:
///  * Report [progress] for each operation that takes non-trivial time, like:
///    * Extracting code samples
///    * Analyzing code samples
///  * Report a [section] for each file being processed.
///  * Report a sub-[section] for each code-sample.
///  * Report an issue for each issue found in a code sample.
abstract class Reporter {
  /// Display a progress bar while [fn] is running.
  ///
  /// Implementer may ignore this if progress bars aren't suitable in the output
  /// format (for example when outputting to JSON or if no terminal is attached).
  Future<void> progress(String message, Future<void> Function() fn);

  /// Report source file found.
  ///
  /// All [sample] invocations following a [file] invocation belongs to said file.
  ///
  /// Implementer may choose to ignore this, or only display it when there is
  /// relevant messages reported for the given file.
  void file(String filename);

  /// Report code sample found in documentation comment.
  ///
  /// This will always be followed by an invocation of [generated].
  /// All [issue] invocations to [issue] following an invocation of [sample] are
  /// issues that belong to said code sample.
  ///
  /// Implementer may choose to print these, simply count them or fully ignore
  /// invocations of this method.
  void sample(SourceSpan span, String source);

  /// Report generated code for code sample [source] identified by [span].
  ///
  /// This is exclusively intended for debugging purposes.
  void generated(SourceSpan span, String source, String generatedFileName, String generatedSource);

  /// Report an issue.
  void issue(SourceSpan span, String message, Severity severity);

  /// Report a debug message.
  void debug(String message);

  /// Report a informational message.
  void info(String message);

  /// Report that analysis is complete, no more messages will be reported.
  ///
  /// Calling any method on a [Reporter] after calling [summary] must
  /// throw [StateError]. Summary is always the last method called.
  ///
  /// Implementer ignore this invocation, or use it to print a summary
  /// of the results.
  void summary({
    required int filesFound,
    required int samplesFound,
    required int issuesFound,
    required Duration elapsed,
  });
}
```

Then we can implement different reporters... We can have a compact reporter that looks like:
```dart

/// A [Reporter] that does display anything.
class SilentReporter extends Reporter {
  @override
  Future<void> progress(String message, Future<void> Function() fn) => fn();

  @override
  void file(String filename) { /* ignore */ }

  @override
  void sample(SourceSpan span, String source) { /* ignore */ }

  @override
  void generated(SourceSpan span, String source, String generatedFileName, String generatedSource) { /* ignore */ }

  @override
  void issue(SourceSpan span, String message, Severity severity) { /* ignore */ }

  @override
  void debug(String message) { /* ignore */ }

  @override
  void info(String message) { /* ignore */ }

  @override
  void summary({
    required int filesFound,
    required int samplesFound,
    required int issuesFound,
    required Duration elapsed,
  }) { /* ignore */ }
}

final class CompactReporter extends SilentReporter {
  final StringSink _sink;

  CompactReporter([_sink = stdout])

  void issue(SourceSpan span, String message, Severity severity) {
    _sink.writeln(span.message(message));
  }

  void info(String message) => print(message);

  void summary({
    required int filesFound,
    required int samplesFound,
    required int issuesFound,
    required Duration elapsed,
  }) {
    if (issuesFound > 0) {
      _sink.writeln('FAILED, $issuesFound issues found (Found $samplesFound code samples across $filesFound files).');
    } else {
      _sink.writeln('SUCCESS, no issues found (Found $samplesFound code samples across $filesFound files).');
    }
  }
}
```

And, I'm sure it's not hard to make a `VerboseReporter` or an `ExtendedReporter` or an `TerminalReporter` (which uses ANSI codes to print colors, bold and underline, maybe uses [cli_util](https://pub.dev/documentation/cli_util/latest/cli_logging/AnsiProgress-class.html) to create a progress bar).
And maybe we can make a `GithubReporter` that makes extra pretty messages for github action logs.

Reply:
If we want to be able to also log messages we could also add a `Logger` property on the `Reporter` or make `class Reporter extends Logger`.

Then we'd have both `Reporter` and `Logger`. I can see how in some cases it might be nice to have the ability to just log messages. And a reporter can then just decide if it wants to show such log messages or not.